### PR TITLE
SAMZA-1651: Samza-sql - Implement GROUP BY SQL operator

### PR DESCRIPTION
**Pull request by atoomula**

### CODE REVIEW

File: samza-sql/src/main/java/org/apache/samza/sql/runner/SamzaSqlApplicationConfig.java [hunk 1]
```diff
@@ -92,6 +95,9 @@
 
   private final List<QueryInfo> queryInfo;
 
+  private final long windowDurationMs;
+  private static final long DEFAULT_WINDOW_DURATION_MS = 300000; // default window duration is 5 mins.
```
Comment (pdu-mn1) [inline samza-sql/src/main/java/org/apache/samza/sql/runner/SamzaSqlApplicationConfig.java]: nit: move constant to top

File: samza-sql/src/main/java/org/apache/samza/sql/runner/SamzaSqlApplicationConfig.java [hunk 2]
```diff
@@ -77,6 +77,9 @@
 
   public static final String CFG_UDF_RESOLVER = "samza.sql.udfResolver";
   public static final String CFG_FMT_UDF_RESOLVER_DOMAIN = "samza.sql.udfResolver.%s.";
+
+  public static final String CFG_WINDOW_DURATION_MS = "samza.sql.window.ms";
```
Comment (pdu-mn1) [inline samza-sql/src/main/java/org/apache/samza/sql/runner/SamzaSqlApplicationConfig.java]: nit: should we prefix this with groupby, ie. "samza.sql.groupby.window.ms", since this is only used by groupBy?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java [hunk 3]
```diff
@@ -96,6 +98,14 @@ public RelNode visit(LogicalJoin join) {
         new JoinTranslator(joinId[0], sourceResolver).translate(join, context);
         return node;
       }
+
+      @Override
+      public RelNode visit(LogicalAggregate aggregate) {
+        RelNode node = super.visit(aggregate);
+        windowId[0] += 2;
```
Comment (pdu-mn1) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java]: nit: can you add a comment on why this is incremented by 2 for each node?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/LogicalAggregateTranslator.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  void translate(final LogicalAggregate aggregate, final TranslatorContext context) {
+    MessageStream<SamzaSqlRelMessage> inputStream = context.getMessageStream(aggregate.getInput().getId());
+
+    Supplier<Long> initialValue = () -> (long) 0;
+    FoldLeftFunction<SamzaSqlRelMessage, Long> foldLeftFn = (m, c) -> c + 1;
+
+    MessageStream<SamzaSqlRelMessage> outputStream =
+        inputStream
```
Comment (pdu-mn1) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/LogicalAggregateTranslator.java]: curious: do we need to expose early triggers for groupby?

File: samza-sql/src/test/java/org/apache/samza/sql/system/TestAvroSystemFactory.java.orig [hunk 5]
```diff
@@ -0,0 +1,299 @@
+/*
```
Comment (pdu-mn1) [inline samza-sql/src/test/java/org/apache/samza/sql/system/TestAvroSystemFactory.java.orig]: nit: This should not be part of the PR right?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/LogicalAggregateTranslator.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+  LogicalAggregateTranslator(int windowId) {
+    this.windowId = windowId;
+  }
+
+  void translate(final LogicalAggregate aggregate, final TranslatorContext context) {
+    MessageStream<SamzaSqlRelMessage> inputStream = context.getMessageStream(aggregate.getInput().getId());
+
+    Supplier<Long> initialValue = () -> (long) 0;
+    FoldLeftFunction<SamzaSqlRelMessage, Long> foldLeftFn = (m, c) -> c + 1;
```
Comment (srinipunuru) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/LogicalAggregateTranslator.java]: This foldLeftFun is specifically for count right? Can you call it as countFn?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java [hunk 7]
```diff
@@ -65,6 +66,7 @@ public void translate(SamzaSqlQueryParser.QueryInfo queryInfo, StreamGraph strea
     final TranslatorContext context = new TranslatorContext(streamGraph, relRoot, executionContext);
     final RelNode node = relRoot.project();
     final int[] joinId = new int[1];
+    final int[] windowId = new int[1];
```
Comment (srinipunuru) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java]: Looks like both windowId and joinId is used only within in the anonymous fn. I think you can move them  inside the anonymous class RelShuttleImpl, that ways you could convert them into a regular integer rather than a single valued array?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java (hunk 3 above)
Comment (srinipunuru) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java]: Can you add a comment on why this needs to be increased by 2?

File: samza-sql/src/test/java/org/apache/samza/sql/system/TestAvroSystemFactory.java.orig [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
```
Comment (srinipunuru) [inline samza-sql/src/test/java/org/apache/samza/sql/system/TestAvroSystemFactory.java.orig]: Is this file checked in by mistake? it's called .java.orig
